### PR TITLE
Fix text encoding handling in C++ DXF importer

### PR DESCRIPTION
**Pull request by kpemartin**
This addresses part of [Issue #8704](https://github.com/FreeCAD/FreeCAD/issues/8704), relating to the C++ DXF importer. The Issue also relates to the legacy importer, where [a separate PR](https://github.com/yorikvanhavre/Draft-dxf-importer/pull/24) has been placed in the appropriate repository.

The fix analyzes the variables at the start of the DXF file to determine the appropriate encoding for text which occurs later in the file, and applies that encoding as required.

- [X]  Your Pull Request meets the requirements outlined in section 5 of [CONTRIBUTING.md](https://github.com/FreeCAD/FreeCAD/blob/master/CONTRIBUTING.md) for a Valid PR

### CODE REVIEW

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8862">pipeline status</a> for feature branch [PR_8862](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8862). Pipeline [803745133](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/803745133) was triggered at [e79c85d](https://github.com/FreeCAD/FreeCAD/pull/8862/commits/e79c85ddfe16ab35291f68bbfdfa73a3abcf3af3). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by luzpaz** [comment]
CI result:  
```
[ 69%] Building CXX object src/Mod/Import/App/CMakeFiles/Import.dir/dxf/dxf.cpp.o
[ 69%] Built target Robot
../../../../../src/Mod/Import/App/dxf/dxf.cpp: In member function 'bool CDxfRead::ReadDWGCodePage()':
../../../../../src/Mod/Import/App/dxf/dxf.cpp:3309:18: error: '_strdup' was not declared in this scope; did you mean 'strdup'?
 3309 |     m_CodePage = _strdup(m_str);
      |                  ^~~~~~~
      |                  strdup
../../../../../src/Mod/Import/App/dxf/dxf.cpp: In member function 'bool CDxfRead::ResolveEncoding()':
../../../../../src/Mod/Import/App/dxf/dxf.cpp:3317:22: error: '_strdup' was not declared in this scope; did you mean 'strdup'?
 3317 |         m_encoding = _strdup("utf_8");
      |                      ^~~~~~~
      |                      strdup
../../../../../src/Mod/Import/App/dxf/dxf.cpp:3322:22: error: '_strdup' was not declared in this scope; did you mean 'strdup'?
 3322 |         m_encoding = _strdup("cp1252");
      |                      ^~~~~~~
      |                      strdup
../../../../../src/Mod/Import/App/dxf/dxf.cpp:3338:26: error: '_strdup' was not declared in this scope; did you mean 'strdup'?
 3338 |             m_encoding = _strdup(m_CodePage);
      |                          ^~~~~~~
      |                          strdup
make[2]: *** [src/Mod/Import/App/CMakeFiles/Import.dir/build.make:202: src/Mod/Import/App/CMakeFiles/Import.dir/dxf/dxf.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:4227: src/Mod/Import/App/CMakeFiles/Import.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

**Comment by xtemp09** [comment]
This code is Windows-only, while the compilation takes place on Ubuntu.

**Comment by chennes** [comment]
> Why does FreeCAD not have Windows CI?

It's too slow right now: apparently it takes something like two hours to run because there's no caching mechanism (like ccache on Linux).

Regardless, the code must compile on Linux, what do you mean when you say it is "Windows only"?

**Comment by xtemp09** [comment]
The code utilizes `_strdup`, which is a function from the C++ extension present in MSVC. The commit has to be rewritten or amended before compilation on Linux or with gcc.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8862">pipeline status</a> for feature branch [PR_8862](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8862). Pipeline [804471659](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/804471659) was triggered at [d9eca55](https://github.com/FreeCAD/FreeCAD/pull/8862/commits/d9eca55a66abacb0482c0446c345a8211c065f49). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by xtemp09** [comment]
In this case, I think, it should check the compiler. If it's MSVC/clang-x86_64-pc-windows-msvc or mingw-w64-gcc/clang-x86_64-w64-windows-gnu.

**Comment by WandererFan** [comment]
Using std::string instead of C strings would avoid the _strdup thing altogether.

Using STL container std::vector<std::string> VersionNames( { "AC1006", ...} )
would avoid the awkward C style determination of the number of elements - VersionNames.size() is much clearer than i = arraybytecount / elementbytecount.

**Comment by WandererFan** [comment]
From a design standpoint, I might separate the code that determines the version/encoding from the dxf handler and just have a parameter to dxf that specifies utf8 vs ascii_with_odd_extensions.

I might even consider pre-processing the pre-2007 file into standard utf8, but I'd have to see how long that pre-processing takes before making a final decision.

**Comment by kpemartin** [comment]
I've updated the code to avoid using strdup.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8862">pipeline status</a> for feature branch [PR_8862](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8862). Pipeline [804619439](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/804619439) was triggered at [b72ed80](https://github.com/FreeCAD/FreeCAD/pull/8862/commits/b72ed80309802346159dee7144a60ff5359f4c13). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8862">pipeline status</a> for feature branch [PR_8862](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8862). Pipeline [804629587](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/804629587) was triggered at [eb6310b](https://github.com/FreeCAD/FreeCAD/pull/8862/commits/eb6310b65f713a3116c4c330ce24dcdb160c98d8). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8862">pipeline status</a> for feature branch [PR_8862](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8862). Pipeline [804776540](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/804776540) was triggered at [f2c197e](https://github.com/FreeCAD/FreeCAD/pull/8862/commits/f2c197e7bb49dc1a1a4cf9ef0a53d2a3808fc94f). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8862">pipeline status</a> for feature branch [PR_8862](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8862). Pipeline [804794328](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/804794328) was triggered at [8d2236a](https://github.com/FreeCAD/FreeCAD/pull/8862/commits/8d2236a8300ba9f9d66479dbfcc7eab8297396c2). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8862">pipeline status</a> for feature branch [PR_8862](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8862). Pipeline [804803668](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/804803668) was triggered at [45ee407](https://github.com/FreeCAD/FreeCAD/pull/8862/commits/45ee407645c147147656ec63cec9da5674516493). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Review by WandererFan — APPROVED** [review]
strcmp? strncmp?  that takes me way back. ;)

        if (m_CodePage->substr(0, 5) == "ansi_")  &&
            m_CodePage->substr(0, 7) != "ansi_x3) {
            // m_CodePage starts with "ansi_" but not with "ansi_x3"

**Comment by kpemartin** [comment]
It's been a while since I've done any serious C++ coding, I guess I set my mind's Wayback Machine a little too far :-)
Even when were using C++ we didn't make much use of the std:: library as we had our own support code that we had carried forward from C, so the <strings.h> stuff was the norm.

Still, my (even earlier) coding instincts rail at how much happens under the hood when you code `m_CodePage->substr(0, 5) == "ansi_"` compared to `strncmp(m_CodePage->c_str(), "ansi_", 5) == 0`. If `std::string` had some sort of "startsWith" method that would avoid having to code the `5` I would have gone with that.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8862">pipeline status</a> for feature branch [PR_8862](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8862). Pipeline [815673323](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/815673323) was triggered at [068acc8](https://github.com/FreeCAD/FreeCAD/pull/8862/commits/068acc84ac7777bc85da4c83db3b700edb06404f). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by kpemartin** [comment]
@yorikvanhavre you're listed as a pending reviewer on this PR. Again, my unfamiliarity with GitHub, is this what is holding up this PR? Does this branch of FreeCAD use automatic or manual PR merging?

I accidentally put my commits on my main 'master' branch and submitted that for this PR. I'd like to get back to having my master branch track the origin code, but every time I update this branch it triggers a new set of automatic checks. I might be able to rename this branch and make myself a new 'master' but that might mess up this PR or give me a mess to clean up on my local computer, so I'd rather just wait until this PR is closed, assuming it won't take too long.

**Pull request closed without merging**
